### PR TITLE
Lesson 5

### PR DESCRIPTION
**Pull request by Teck1**
Initial pull for exceptions exercise.

### CODE REVIEW

File: students/ChrisH/lesson04/mailroom.py [hunk 1]
```diff
@@ -88,7 +88,12 @@ def send_thank_you():
             donor_data[name] = {'first_name': name_sp[0], 'last_name': ' '.join(name_sp[1:]), 'donations': []}
         break
 
-    amount = input("Enter a donation amount for {} : ".format(name))
+    while True:
+        amount = input("Enter a donation amount for {} : ".format(name))
+        if float(amount) <= 0:
+            print('Amount donated must be a positive number.')
```
Comment (mdgregor) [inline students/ChrisH/lesson04/mailroom.py]: :)

File: students/ChrisH/lesson05/mailroom.py [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+        Runs through donor data structure, generates a thank you letter for each and saves it to
+        the current working directory with in a date+donorname.txt file
+        :return: None
+        """
+    for donor in donor_data:
+        print(f'Generating letter for {get_donor_fullname(donor)}')
+        now = time.localtime()
+        f_name = f"{now.tm_year}{now.tm_mon:02}{now.tm_mday:02}_"
+        f_name += get_donor_fullname(donor).replace(" ", "_") + ".txt"
+        file_out = open(f_name, 'w')
```
Comment (mdgregor) [inline students/ChrisH/lesson05/mailroom.py]: Can you use a `with` statement here instead?

File: students/ChrisH/lesson05/mailroom.py [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+        print(f"{get_donor_fullname(donor):{name_max}}  ", end='')
+        print(f"$ {sum(dons):>10.2f}   {len(dons):>9}  ${sum(dons)/len(dons):>12.2f}")
+
+
+def nul():
+    pass
+
+
+if __name__ == "__main__":
+    menu_functions = {0: ['nul', nul],                      # could also use 'lambda: None' instead of nul
```
Comment (mdgregor) [inline students/ChrisH/lesson05/mailroom.py]: why have a `0` key if its never used?

File: students/ChrisH/lesson05/mailroom.py [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+    Prints the main user menu & retrieves user selection.
+    :param: a menu dictionary, key as numbered list.
+            Dict key 0 should be a null function (or text for bad choice)
+            Dict value is a list with the following entries:
+                [0] = Text for user prompt
+                [1] = Function that should be run for that choice
+    :return: the function corresponding to the user's selection
+    """
+    print("\nPlease choose one of the following options:")
+    for n in range(1, len(menu_data)):                     # Start at item 1, item 0 should be null function
```
Comment (mdgregor) [inline students/ChrisH/lesson05/mailroom.py]: Change this to `for n in range(1, len(menu_data) + 1):  ` and you don't need the `0` menu option

**Comment by mdgregor** [comment]
In regards to the menu, i think we could simplify it a bit while keeping your same idea:

```if __name__ == "__main__":
    menu_functions = {
        'Send a Thank You': send_thank_you,
        'Print a report': print_report,
        'Send letters to everyone': send_letters_all,
        'Quit': quit
    }

    while True:
        menu(menu_functions)()```

```
print("\nPlease choose one of the following options:")
    for index, key in enumerate(menu_data.keys()):
        print(f"{index + 1}) {key}")
    choice = int(input("> "))
    for index, value in enumerate(menu_data.values()):
        if choice == index + 1:
            return value```

**Comment by Teck1** [comment]
I like your suggestion for the menu changes. It does simplify the menu() function. It eliminates the menu 0 option, and removes the need for option numbering by using an index value. I tried it, it seemed to work well, but I then had second thoughts.

Doesn't the dictionary structure forego the capability of having an ordered menu? e.g. the Quit menu option could be presented anywhere in the list, according to how the dictionary stores the key/value pair. It could look odd being the first menu choice. 

Maybe a list of tuples would be better? It keeps the order, doesn't require the index number, and pairs text with a function name. Hmm, trying that. ... And, I think I like the list of tuples better! I'll commit it with my next revision.

File: students/ChrisH/lesson05/mailroom.py (hunk 2 above)
Comment (Teck1) [reply students/ChrisH/lesson05/mailroom.py]: Oh, yeah! Forgot that trick for a moment. Fixed.

**Comment by mdgregor** [comment]
Yes you are correct, but in python 3.6 they implemented order preservation, though they also state it shouldn't be relied upon.

```
New dict implementation
The dict type now uses a “compact” representation based on a proposal by Raymond Hettinger which was first implemented by PyPy. The memory usage of the new dict() is between 20% and 25% smaller compared to Python 3.5.

The order-preserving aspect of this new implementation is considered an implementation detail and should not be relied upon (this may change in the future, but it is desired to have this new dict implementation in the language for a few releases before changing the language spec to mandate order-preserving semantics for all current and future Python implementations; this also helps preserve backwards-compatibility with older versions of the language where random iteration order is still in effect, e.g. Python 3.5).```
https://docs.python.org/3/whatsnew/3.6.html

**Comment by Teck1** [comment]
This is my revision, it keeps a lot of the cleanup you made, but changes the structure to tuple instead of dictionary. I guess it really only buys me not having to loop the dictionary twice (keys, values), but also costs me a check on the user choice, because - slicing rules! If the user chooses zero, they'll get the [-1] item. 
```
    menu_functions = [
        ('Send a Thank You', send_thank_you),
        ('Print a report', print_report),
        ('Send letters to everyone', send_letters_all),
        ('Quit', quit),
    ]
    while True:
        try:
            menu(menu_functions)()
        except TypeError:
            continue
        except ValueError:
            continue
```
```
    print("\nPlease choose one of the following options:")
    for index, menu_item in enumerate(menu_data):   # Prints the menu user text
        print(f"{index + 1}) {menu_item[0]}")

    choice = int(input("> ")) - 1

    if choice in range(len(menu_data)):             # Ensure that option chosen is within menu range
        return menu_data[choice][1]                 # this handles choosing 0, which would return menu_data[-1][1]
    return None
```

**Pull request merged**